File: ios/Wearther/Wearther/Shared/FitCopy.swift
import Foundation

enum FitConfidence: String, Equatable {
    case confident = "Confident"
    case bringBackup = "Bring backup"
    case rainRisk = "Rain risk"
    case eveningDrop = "Evening drop"
    case tunedForYou = "Tuned for you"

    var systemImage: String {
        switch self {
        case .confident: return "checkmark"
        case .bringBackup: return "bag"
        case .rainRisk: return "cloud.rain"
        case .eveningDrop: return "moon.stars"
        case .tunedForYou: return "slider.horizontal.3"
        }
    }
}

enum FitCopy {
    static func formatTitle(_ outfit: OutfitRecommendation) -> String {
        let jacketOptions: String.CompareOptions = [.regularExpression, .caseInsensitive]
        let hasJacket = outfit.items.contains {
            $0.range(of: "jacket|coat|bring", options: jacketOptions) != nil
        }
        if hasJacket || outfit.title.range(of: "jacket|coat|bring", options: jacketOptions) != nil {
            return outfit.title
        }
        if outfit.title.range(of: "no jacket", options: jacketOptions) != nil {
            return outfit.title
        }
        return "\(outfit.title), No Jacket"
    }

    static func confidence(
        outfit: OutfitRecommendation,
        weather: WeatherData,
        comfort: ComfortPreference
    ) -> FitConfidence {
        if comfort.feedbackCount > 0 || comfort.lastFeedback != nil {
            return .tunedForYou
        }
        if weather.precipitationChance >= 45 {
            return .rainRisk
        }
        if outfit.bringLater != nil {
            return .eveningDrop
        }
        let tip = packTip(outfit: outfit, weather: weather)
        if tip.value.lowercased() != "travel light" {
            return .bringBackup
        }
        return .confident
    }

    /// Back-compat for widget / older call sites.
    static func confidenceLabel(_ comfort: ComfortPreference) -> String {
        if comfort.feedbackCount > 0 || comfort.lastFeedback != nil {
            return FitConfidence.tunedForYou.rawValue
        }
        return FitConfidence.confident.rawValue
    }

    static func confidenceLabel(
        outfit: OutfitRecommendation,
        weather: WeatherData,
        comfort: ComfortPreference
    ) -> String {
        confidence(outfit: outfit, weather: weather, comfort: comfort).rawValue
    }

    static func packTip(outfit: OutfitRecommendation, weather: WeatherData) -> (label: String, value: String) {
        if let later = outfit.bringLater {
            let short = later
                .replacingOccurrences(
                    of: #"^bring\s+"#,
                    with: "",
                    options: [.regularExpression, .caseInsensitive]
                )
                .trimmingCharacters(in: CharacterSet(charactersIn: "."))
            return ("Pack", short.prefix(1).uppercased() + short.dropFirst())
        }
        if weather.precipitationChance >= 40 {
            return ("Pack", "Rain layer")
        }
        if weather.high - weather.low >= 12 {
            return ("Pack", "Light layer")
        }
        return ("Pack", "Travel light")
    }

    /// First-class pack lane cards (empty when nothing special to pack).
    static func packLaneItems(outfit: OutfitRecommendation, weather: WeatherData) -> [String] {
        var items: [String] = []

        if let later = outfit.bringLater {
            let short = later
                .replacingOccurrences(
                    of: #"^bring\s+(a\s+)?"#,
                    with: "",
                    options: [.regularExpression, .caseInsensitive]
                )
                .trimmingCharacters(in: CharacterSet(charactersIn: "."))
            let titled = short.prefix(1).uppercased() + short.dropFirst()
            items.append(titled)
        }

        if weather.precipitationChance >= 40 {
            let rain = "Rain shell"
            if !items.contains(where: { $0.localizedCaseInsensitiveContains("rain") }) {
                items.append(rain)
            }
        }

        if weather.high - weather.low >= 12,
           !items.contains(where: { $0.localizedCaseInsensitiveContains("layer") || $0.localizedCaseInsensitiveContains("jacket") }) {
            items.append("Light layer")
        }

        return items
    }

    static func shortExplanation(_ outfit: OutfitRecommendation) -> String {
        let first = outfit.explanation
            .split(separator: ".")
            .first
            .map(String.init)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard let first, !first.isEmpty else { return outfit.explanation }
        return first.hasSuffix(".") ? first : "\(first)."
    }

    static func whyDetail(weather: WeatherData, outfit: OutfitRecommendation) -> String {
        if let detail = outfit.whyDetail, !detail.isEmpty {
            return detail
        }

        var parts: [String] = []
        parts.append("Feels like \(weather.feelsLike)°")

        if weather.humidity >= 70 {
            parts.append("humidity is high (\(weather.humidity)%)")
        } else if weather.humidity <= 35 {
            parts.append("humidity is low (\(weather.humidity)%)")
        } else {
            parts.append("humidity is moderate (\(weather.humidity)%)")
        }

        if weather.windSpeed >= 12 {
            parts.append("wind is strong (\(weather.windSpeed) mph)")
        } else if weather.windSpeed >= 8 {
            parts.append("wind is breezy (\(weather.windSpeed) mph)")
        } else {
            parts.append("wind is mild (\(weather.windSpeed) mph)")
        }

        if weather.precipitationChance >= 45 {
            parts.append("rain chance is \(weather.precipitationChance)%")
        }

        if let hourlyDrop = eveningDropDegrees(weather: weather) {
            if hourlyDrop >= 10 {
                parts.append("evening drops about \(hourlyDrop)°")
            } else {
                parts.append("evening only drops \(hourlyDrop)°")
            }
        }

        return parts.joined(separator: ", ") + "."
    }

    static func feedbackResponse(_ feedback: ComfortFeedback) -> String {
        switch feedback {
        case .tooCold:
            return "Got it. Tomorrow will lean warmer."
        case .perfect:
            return "Nice. Keeping this baseline."
        case .tooHot:
            return "Got it. Tomorrow will lighten up."
        }
    }

    static func notificationBody(
        locationName: String,
        weather: WeatherData,
        outfit: OutfitRecommendation
    ) -> String {
        let title = formatTitle(outfit)
        let breeze = weather.windSpeed >= 10 ? "breezy" : weather.condition.lowercased()
        var body = "\(locationName) is \(breeze). \(title)."
        let packs = packLaneItems(outfit: outfit, weather: weather)
        if let first = packs.first {
            let when: String
            if let hour = laterPackHour(weather: weather) {
                when = " after \(hour)"
            } else {
                when = ""
            }
            body += " Pack \(first.lowercased())\(when)."
        }
        return body
    }

    private static func eveningDropDegrees(weather: WeatherData) -> Int? {
        guard !weather.hourly.isEmpty else {
            let span = weather.high - weather.low
            return span > 0 ? span : nil
        }
        let coldest = weather.hourly.map(\.feelsLike).min() ?? weather.feelsLike
        return max(0, weather.feelsLike - coldest)
    }

    private static func laterPackHour(weather: WeatherData) -> String? {
        guard let coldest = weather.hourly.min(by: { $0.feelsLike < $1.feelsLike }) else {
            return "6 PM"
        }
        guard let date = parseTime(coldest.time) else { return "6 PM" }
        let hour = Calendar.current.component(.hour, from: date)
        let formatter = DateFormatter()
        formatter.dateFormat = "h a"
        var comps = DateComponents()
        comps.hour = hour
        guard let display = Calendar.current.date(from: comps) else { return "6 PM" }
        return formatter.string(from: display)
    }

    private static func parseTime(_ iso: String) -> Date? {
        if let d = ISO8601DateFormatter().date(from: iso) { return d }
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm"
        return formatter.date(from: String(iso.prefix(16)))
    }
}

File: src/lib/fitCopy.ts
import type {
  ComfortFeedback,
  ComfortPreference,
  FitConfidence,
  OutfitRecommendation,
} from "@/types/outfit";
import { getRemoteConfig } from "./remoteConfig";
import type { WeatherData } from "@/types/weather";

export function formatFitTitle(outfit: OutfitRecommendation): string {
  const hasJacket = outfit.items.some((i) => /jacket|coat|bring/i.test(i));
  if (hasJacket || /jacket|coat|bring/i.test(outfit.title)) return outfit.title;
  if (/no jacket/i.test(outfit.title)) return outfit.title;
  return `${outfit.title}, No Jacket`;
}

export function confidenceForFit(
  outfit: OutfitRecommendation,
  weather: WeatherData,
  comfort: ComfortPreference,
): FitConfidence {
  const { thresholds, copy } = getRemoteConfig();
  if (comfort.feedbackCount > 0 || comfort.lastFeedback) {
    return copy.confidence.tunedForYou as FitConfidence;
  }
  if (weather.precipitationChance >= thresholds.highRainChance) {
    return copy.confidence.rainRisk as FitConfidence;
  }
  if (outfit.bringLater) {
    return copy.confidence.eveningDrop as FitConfidence;
  }
  const tip = packTip(outfit, weather);
  if (tip.toLowerCase() !== copy.packTips.travelLight.toLowerCase()) {
    return copy.confidence.bringBackup as FitConfidence;
  }
  return copy.confidence.confident as FitConfidence;
}

export function packTip(outfit: OutfitRecommendation, weather: WeatherData): string {
  const { thresholds, copy } = getRemoteConfig();
  if (outfit.bringLater) {
    const short = outfit.bringLater
      .replace(/^bring\s+/i, "")
      .replace(/\.$/, "")
      .trim();
    return short.charAt(0).toUpperCase() + short.slice(1);
  }
  if (weather.precipitationChance >= thresholds.packRainChance) {
    return copy.packTips.rainLayer;
  }
  if (weather.high - weather.low >= thresholds.diurnalSpanF) {
    return copy.packTips.lightLayer;
  }
  return copy.packTips.travelLight;
}

export function packLaneItems(
  outfit: OutfitRecommendation,
  weather: WeatherData,
): string[] {
  const { thresholds, copy } = getRemoteConfig();
  const items: string[] = [];

  if (outfit.bringLater) {
    const short = outfit.bringLater
      .replace(/^bring\s+(a\s+)?/i, "")
      .replace(/\.$/, "")
      .trim();
    items.push(short.charAt(0).toUpperCase() + short.slice(1));
  }

  if (weather.precipitationChance >= thresholds.packRainChance) {
    if (!items.some((i) => /rain/i.test(i))) items.push("Rain shell");
  }

  if (
    weather.high - weather.low >= thresholds.diurnalSpanF &&
    !items.some((i) => /layer|jacket/i.test(i))
  ) {
    items.push(copy.packTips.lightLayer);
  }

  return items;
}

export function shortExplanation(outfit: OutfitRecommendation): string {
  const first = outfit.explanation.split(".")[0]?.trim();
  if (!first) return outfit.explanation;
  return first.endsWith(".") ? first : `${first}.`;
}

export function whyDetail(
  weather: WeatherData,
  outfit: OutfitRecommendation,
): string {
  if (outfit.whyDetail) return outfit.whyDetail;

  const t = getRemoteConfig().thresholds;
  const parts: string[] = [`Feels like ${weather.feelsLike}°`];

  if (weather.humidity >= t.highHumidity) {
    parts.push(`humidity is high (${weather.humidity}%)`);
  } else if (weather.humidity <= t.lowHumidity) {
    parts.push(`humidity is low (${weather.humidity}%)`);
  } else {
    parts.push(`humidity is moderate (${weather.humidity}%)`);
  }

  if (weather.windSpeed >= t.strongWindMph) {
    parts.push(`wind is strong (${weather.windSpeed} mph)`);
  } else if (weather.windSpeed >= 8) {
    parts.push(`wind is breezy (${weather.windSpeed} mph)`);
  } else {
    parts.push(`wind is mild (${weather.windSpeed} mph)`);
  }

  if (weather.precipitationChance >= t.highRainChance) {
    parts.push(`rain chance is ${weather.precipitationChance}%`);
  }

  if (weather.hourly.length) {
    const coldest = Math.min(...weather.hourly.map((h) => h.feelsLike));
    const drop = Math.max(0, weather.feelsLike - coldest);
    parts.push(
      drop >= t.significantDropF
        ? `evening drops about ${drop}°`
        : `evening only drops ${drop}°`,
    );
  }

  return `${parts.join(", ")}.`;
}

export function feedbackResponse(feedback: ComfortFeedback): string {
  const { feedback: copy } = getRemoteConfig().copy;
  if (feedback === "too_cold") return copy.too_cold;
  if (feedback === "too_hot") return copy.too_hot;
  return copy.perfect;
}

export function updatedLabel(fetchedAt: string): string {
  const date = new Date(fetchedAt);
  if (Number.isNaN(date.getTime())) return "Updated just now";
  return `Updated ${date.toLocaleTimeString([], { hour: "numeric", minute: "2-digit" })}`;
}

export function isWeatherStale(fetchedAt: string): boolean {
  const date = new Date(fetchedAt);
  if (Number.isNaN(date.getTime())) return true;
  const minutes = getRemoteConfig().timing.weatherStaleMinutes;
  return Date.now() - date.getTime() > minutes * 60 * 1000;
}
